pkg/k8s/identitybackend/identity.go

```diff
@@ -0,0 +1,418 @@
+// AllocatorKey.GetAsMap (when the key is k8s labels). In the CRD identity case
+// we map the labels directly to the ciliumidentity CRD instance, and
+// kubernetes does not allow ':' in the name of the label.
+func stripK8sPrefix(old map[string]string) map[string]string {
+	fixup := make(map[string]string, len(old))
+	for k, v := range old {
+		k = strings.TrimPrefix(k, "k8s:")
+		fixup[k] = v
+	}
+	return fixup
+}
+
+// AllocateID will create an identity CRD, thus creating the identity for this
+// key-> ID mapping.
+// Note: This does not create a reference to this node to indicate that it is
+// using this identity. That must be done with AcquireReference.
+// Note: the lock field is not supported with the k8s CRD allocator.
+func (c *crdBackend) AllocateID(ctx context.Context, id idpool.ID, key allocator.AllocatorKey) error {
+	identity := &v2.CiliumIdentity{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   id.String(),
+			Labels: stripK8sPrefix(key.GetAsMap()),
+		},
+		Status: v2.IdentityStatus{
+			Nodes: map[string]metav1.Time{
+				c.NodeName: metav1.Now(),
+			},
+		},
+	}
+
+	_, err := c.Client.CiliumV2().CiliumIdentities().Create(identity)
+	return err
+}
+
+func (c *crdBackend) AllocateIDIfLocked(ctx context.Context, id idpool.ID, key allocator.AllocatorKey, lock kvstore.KVLocker) error {
+	return c.AllocateID(ctx, id, key)
+}
+
+// JSONPatch structure based on the RFC 6902
+type JSONPatch struct {
+	OP    string      `json:"op,omitempty"`
+	Path  string      `json:"path,omitempty"`
+	Value interface{} `json:"value"`
+}
+
+// AcquireReference updates the status field of the CRD corresponding to id
+// with this node. This marks that CRD as used by this node, and will stop it
+// being garbage collected.
+// Note: the lock field is not supported with the k8s CRD allocator.
+func (c *crdBackend) AcquireReference(ctx context.Context, id idpool.ID, key allocator.AllocatorKey, lock kvstore.KVLocker) error {
+	identity := c.get(ctx, key)
+	if identity == nil {
+		return fmt.Errorf("identity does not exist")
+	}
+
+	capabilities := k8sversion.Capabilities()
+	identityOps := c.Client.CiliumV2().CiliumIdentities()
+
+	var err error
+	if capabilities.Patch {
+		var patch []byte
+		patch, err = json.Marshal([]JSONPatch{
+			{
+				OP:    "test",
+				Path:  "/status",
+				Value: nil,
+			},
+			{
+				OP:   "add",
+				Path: "/status",
+				Value: v2.IdentityStatus{
+					Nodes: map[string]metav1.Time{
+						c.NodeName: metav1.Now(),
+					},
+				},
+			},
+		})
+		if err != nil {
+			return err
+		}
+
+		_, err = identityOps.Patch(identity.GetName(), k8sTypes.JSONPatchType, patch, "status")
+		if err != nil {
+			patch, err = json.Marshal([]JSONPatch{
+				{
+					OP:    "replace",
+					Path:  "/status/nodes/" + c.NodeName,
+					Value: metav1.Now(),
+				},
+			})
+			if err != nil {
+				return err
+			}
+			_, err = identityOps.Patch(identity.GetName(), k8sTypes.JSONPatchType, patch, "status")
```

Comment:
If this error is non-nil, we will ignore it entirely right? Should we be logging it?

Reply:
I presume this is the fallback case where we fall back to a full update of the CRD entry rather than patching it, so logically seems "ok".

Reply:
yes, it is a fallback but it is still an unexpected error. I'll debug print on error. That won't alarm anyone under normal usage, but might improve a debugger's life one day. A+1 question.